src/lustre/lustreTypeChecker.ml

```diff
@@ -0,0 +1,1379 @@
+  = fun ctx eqn ->
+  Log.log L_debug "checking contract with equation %a" LA.pp_print_contract_item eqn 
+  ; match eqn with
+  | GhostConst _
+  | GhostVar _ ->  R.ok () (* These is already checked while extracting ctx *)
+  | Assume (pos, _, _, e) -> check_type_expr ctx e (Bool pos)
+  | Guarantee (pos, _, _, e) -> check_type_expr ctx e (Bool pos)
+  | Mode (pos, _, reqs, ensures) ->
+     R.seq_ (Lib.list_apply (List.map (check_type_expr ctx)
+                               (List.map (fun (_,_, e) -> e) (reqs @ ensures)))
+               (Bool pos)) 
+  | ContractCall (pos, cname, args, rets) ->
+     R.seq(List.map (infer_type_expr ctx) args)
+     >>= fun ret_tys ->  
+     let ret_ty = if List.length ret_tys = 1
+                  then List.hd ret_tys
+                  else LA.TupleType (pos, ret_tys) in
+     check_type_expr ctx (LA.Call (pos, cname, args))(ret_ty) 
+
+and tc_ctx_of_ty_decl: tc_context -> LA.type_decl -> tc_context tc_result
+  = fun ctx ->
+  function
+  | LA.AliasType (_, i, ty) ->
+     (match ty with
+      | LA.EnumType (pos, n, econsts) ->
+         (match n with
+          | None ->
+             type_error pos "Necessary Enum name not found"
+          | Some ename ->
+             if (List.for_all (fun e -> not (member_ty ctx e)) econsts)
+                && (List.for_all (fun e -> not (member_val ctx e)) econsts)
+             then 
+             R.ok (
+                 List.fold_left union (add_ty_decl ctx ename)
+                   ((List.map ((Lib.flip singleton_ty)
+                                 (LA.UserType (pos, ename))) econsts)
+                    @ Lib.list_apply ((List.map2 (Lib.flip singleton_const)
+                         (List.map (fun i -> LA.Ident (pos, i)) econsts) econsts)) (LA.UserType (pos, ename)))
+               )
+             else
+               type_error pos "Cannot redeclare constants or enums")
+      | _ -> R.ok (add_ty_syn ctx i ty))
+  | LA.FreeType (_, i) -> R.ok (add_ty_decl ctx i)
+
+and tc_ctx_of_node_decl: Lib.position -> tc_context -> LA.node_decl -> tc_context tc_result
+  = fun pos ctx (nname, _, _ , ip, op,_ ,_ ,_) ->
+  Log.log L_debug
+    "Extracting typing context from node declaration: %a"
+    LA.pp_print_ident nname
+  ; if (member_ty ctx nname)
+    then type_error pos ("Duplicate node detected with name: " ^ nname)
+    else build_node_fun_ty pos ctx ip op
+         >>= fun fun_ty -> R.ok (add_ty ctx nname fun_ty)
+(** computes the type signature of node or a function *)
+
+and tc_ctx_of_contract_node_decl: Lib.position -> tc_context
+                                  -> LA.contract_node_decl
+                                  -> tc_context tc_result
+  = fun pos ctx (cname, params, inputs, outputs, contrat) ->
+  Log.log L_debug
+    "Extracting typing context from contract declaration: %a"
+    LA.pp_print_ident cname
+  ; if (member_ty ctx cname)
+    then type_error pos ("Duplicate node detected with name: " ^ cname)
```

Comment:
Change message to `"Contract " ^ cname ^ " is already declared."`

Reply:
fixed